### PR TITLE
 "user/playergroup/groups seed created. Rest still in the works. Will change 'city' => 'address' once updated version is pulled"

### PR DESCRIPTION
**Pull request by Ryu4fly**

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -1,7 +1,52 @@
...
-#   Character.create(name: 'Luke', movie: movies.first)
+require 'faker'
+
+puts "Clearing out the battlefield 🌋..."
+User.destroy_all
+PlayerGroup.destroy_all
+Group.destroy_all
+Preference.destroy_all
+Rulebook.destroy_all
+Genre.destroy_all
```
Comment (Naokimi) [inline db/seeds.rb]: use the following instead:
```
require 'database_cleaner/active_record'
DatabaseCleaner.strategy = :truncation
DatabaseCleaner.clean
```

File: db/seeds.rb [hunk 2]
```diff
@@ -1,7 +1,52 @@
...
+puts "Battlefield Cleared ⛰"
+
+puts "Summoning Gandalf 🧙🏻‍♂️"
+gandalf = User.new
```
Comment (Naokimi) [inline db/seeds.rb]: better to use `gandalf = User.create!()` and the put the lines below inside the parentheses

File: db/seeds.rb [hunk 3]
```diff
@@ -1,7 +1,52 @@
...
+  user.nickname = Faker::Internet.username
+  user.password = '123456'
+  user.city = Faker::Address.city
+  user.save
+  puts "#{user.nickname} from #{user.city} created"
+end
+puts "All NPCs Generated"
+
+puts "Generating Groups..."
+50.times do
```
Comment (Naokimi) [inline db/seeds.rb]: too many groups
you usually want 4-5 players per group so create just 20 of them

File: db/seeds.rb [hunk 4]
```diff
@@ -1,7 +1,52 @@
...
+  group.description = Faker::Company.bs
+  group.city = Faker::Address.city
+  group.gm = User.all.shuffle.first
+  group.save
+  puts "#{group.name} created"
+end
+puts "All Groups Created"
+
+puts "Linking Users to Groups..."
+30.times do
```
Comment (Naokimi) [inline db/seeds.rb:104]: in this way you could have a user belonging to the same group twice
try to think of a better way how to match them or add a guard clause :)

**Review by Naokimi — CHANGES_REQUESTED** [review]

File: db/seeds.rb [hunk 5]
```diff
@@ -1,7 +1,65 @@
...
+  email: 'gandalf@gmail.com',
+  nickname: 'GandalfTheWhite',
+  password: '123456',
+  city: 'Tokyo'
+  )
+puts "#{gandalf.nickname} Summoned: 'You Shall Not Pass! 🧖🏼‍♂️'"
+
+puts "Generating NPCs..."
+40.times do
+  user = User.new
```
Comment (Naokimi) [inline db/seeds.rb]: the `.create` should be used everywhere when possible :P

File: db/seeds.rb [hunk 6]
```diff
@@ -1,7 +1,65 @@
...
+DatabaseCleaner.strategy = :truncation
+DatabaseCleaner.clean
+puts "Battlefield Cleared ⛰"
+
+puts "Summoning Gandalf 🧙🏻‍♂️"
+gandalf = User.create!(
+  email: 'gandalf@gmail.com',
+  nickname: 'GandalfTheWhite',
+  password: '123456',
+  city: 'Tokyo'
```
Comment (Naokimi) [inline db/seeds.rb]: btw, for "city" we are actually looking at -ku for better granularity
so 'Meguro-ku, Tokyo"

File: db/seeds.rb [hunk 7]
```diff
@@ -1,7 +1,65 @@
...
+  )
+puts "#{gandalf.nickname} Summoned: 'You Shall Not Pass! 🧖🏼‍♂️'"
+
+puts "Generating NPCs..."
+40.times do
+  user = User.new
+  user.email = Faker::Internet.email
+  user.nickname = Faker::Internet.username
+  user.password = '123456'
+  user.city = Faker::Address.city
```
Comment (Naokimi) [inline db/seeds.rb]: see above about cities
try to have results that are within the greater tokyo area

File: db/seeds.rb [hunk 8]
```diff
@@ -1,7 +1,65 @@
...
+  puts "#{user.nickname} from #{user.city} created"
+end
+puts "All NPCs Generated"
+
+puts "Generating Groups..."
+20.times do
+  group = Group.new
+  group.name = Faker::App.name
+  group.description = Faker::Company.bs
+  group.city = Faker::Address.city
```
Comment (Naokimi) [inline db/seeds.rb]: see as above

File: db/seeds.rb [hunk 9]
```diff
@@ -1,7 +1,65 @@
...
+# See why this doesnt work...
+private
+
+def randUser
+  User.all.sample
+end
+
+def randGroup
+  Group.all.sample
+end
```
Comment (Naokimi) [inline db/seeds.rb]: ok, there's a lot wrong here...

1. follow ruby conventions for naming (i.e. `rand_user` not `randUser`)
2. there is no point putting it in a method if you are not reusing it
3. writing `private` only works withing the context of a Class or Module
4. you don't have a guarantee that the second time you call rand_group the user won't be a gm

look into either using an iterator, or doing a db query with conditions

**Review by Naokimi — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 9 above)
Comment (Ryu4fly) [reply db/seeds.rb]: This part I was going to mention still in works and testing myself (this morning) shouldnt have added it to the commit.
Will follow your instructions

File: db/seeds.rb [hunk 10]
```diff
@@ -1,7 +1,67 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the bin/rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+require_relative 'data/custom_data.rb'
```
Comment (Naokimi) [inline db/seeds.rb]: nice

File: db/data/custom_data.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
+# Data for Address
+WARDS = [ 'Adachi', 'Arakawa', 'Bunkyo', 'Chiyoda', 'Chuo', 'Edogawa', 'Itabashi', 'Katsushika', 'Kita', 'Koto', 'Meguro', 'Minato', 'Nakano', 'Nerima', 'Ota', 'Setagaya', 'Shibuya', 'Shinagawa', 'Shinjuku', 'Suginami', 'Sumida', 'Taito', 'Toshima' ]
```
Comment (Naokimi) [inline db/data/custom_data.rb]: can you please add also wards/cities of prefectures surrounding tokyo? (kanagawa, chiba...)

File: db/test.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+require 'nokogiri'
+
+ingredient = 'chocolate'
+url = "https://rpggeek.com/browse/rpggenre"
+
+html_file = open(url).read
+html_doc = Nokogiri::HTML(html_file)
+
+
+puts html_doc
```
Comment (Naokimi) [inline db/test.rb]: what's this for?

File: db/seeds.rb [hunk 13]
```diff
@@ -1,7 +1,67 @@
...
+  p_group = PlayerGroup.create(
+    user: user,
+    group: group
+    )
+end
+puts "We have Player Groups 🕺"
+
+puts "Generating Genres..."
+RPG_GENRES.each_key do |genre|
+  i = 0
```
Comment (Naokimi) [inline db/seeds.rb]: I think you can do `RPR_GENRES.each_with_index.each_key` to avoid having to write the `i = 0`

File: db/seeds.rb [hunk 14]
```diff
@@ -1,7 +1,67 @@
...
+  puts "#{i += 1}. #{group.name} created"
+end
+puts "All Groups Created"
+
+puts "Linking Users to Groups..."
+30.times do
+  user = User.all.sample
+  group = Group.all.sample
+  group = Group.all.sample if group.gm == user
+  group = Group.all.sample if group.users.count == 5
```
Comment (Naokimi) [inline db/seeds.rb]: this stil isn't going to guarantee a correct group
try using a `while` loop

**Review by Naokimi — CHANGES_REQUESTED** [review]

File: db/seeds.rb (hunk 14 above)
Comment (Ryu4fly) [reply db/seeds.rb]: I am guessing line 50 contradicts line 49??

File: db/test.rb (hunk 12 above)
Comment (Ryu4fly) [reply db/test.rb]: removed... 😅

File: db/data/custom_data.rb (hunk 11 above)
Comment (Ryu4fly) [reply db/data/custom_data.rb]: on it

File: db/seeds.rb (hunk 14 above)
Comment (Naokimi) [reply db/seeds.rb]: not just that, line 49 can return the same group
the chance is small but it's possible
if you want to be 100% sure you need to keep sampling

Comment (Ryu4fly) [reply db/seeds.rb]: ah, gotcha!

Will think of way to write it with `while`

File: db/data/genres_data.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
+# Data for Genres
+RPG_GENRES = [
+  {
+    name: 'Post Apocalypse',
+    description: 'Material set in the wake of the destruction of civilization as we know it. Perhaps caused by human self-destruction or invoked by external unearthly forces, humanity has been reduced to a subsistence state, trying to survive the chaos that remains'
```
Comment (Naokimi) [inline db/data/genres_data.rb]: instead of 'material' I think a more appropriate term would be 'games'

File: db/data/rulebooks_data.rb [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  {
+    id: 17,
+    title: 'Warhammer Fantasy Roleplay (4e)',
+    genre: 'Fantasy'
+  },
+  {
+    id: 20,
+    title: 'Dungeons & Dragons (3e)',
+    genre: 'Fantasy'
+  }
```
Comment (Naokimi) [inline db/data/rulebooks_data.rb]: missing descriptions here which AFAIK would break your seeds

File: db/seeds.rb [hunk 17]
```diff
@@ -1,7 +1,146 @@
...
+          city_or_ward = prefecture.keys[rand(2..3)]
+          city_or_ward == :ward ? "#{prefecture[city_or_ward].sample}-ku" : "#{prefecture[city_or_ward].sample}-shi"
+          else
+            "#{prefecture[:city].sample}-shi"
+         end
+
+  group = Group.create(
+    name: Faker::App.name,
+    description: Faker::Company.bs,
+    city: "#{city}, Tokyo",
```
Comment (Naokimi) [inline db/seeds.rb]: city: "#{city}, #{prefecture[:prefecture]}"

File: db/seeds.rb [hunk 18]
```diff
@@ -1,7 +1,146 @@
...
+Preference.create(
+  user: gandalf,
+  genre: Genre.all.sample
+  )
+
+50.times do
+  Preference.create(
+    user: User.all.sample,
+    genre: Genre.all.sample
+    )
```
Comment (Naokimi) [inline db/seeds.rb]: preferences can't be duplicated, you'll need to check for that

File: db/seeds.rb [hunk 19]
```diff
@@ -1,7 +1,146 @@
...
+end
+
+puts "They decided"
+
+puts "Create the Rulebooks..."
+BOOK_TITLES.each_with_index do |book, index|
+  genre = Genre.where(name: book[:genre]).first
```
Comment (Naokimi) [inline db/seeds.rb]: use `.find_by` instead of `.where().first`

**Review by Naokimi — CHANGES_REQUESTED** [review]
not sure if you officially requested a review (I keep getting notifications when you push to this PR), but before requesting next time please make sure to run the seeds 2 or 3 times and make sure nothing breaks and everything was created :)

also you should be using everywhere `.create!` instead of `.create` because otherwise you won't know when something fails

File: db/data/rulebooks_data.rb (hunk 16 above)
Comment (Ryu4fly) [reply db/data/rulebooks_data.rb]: strange, I had descriptions... will recommit after changes made~

File: db/seeds.rb [hunk 20]
```diff
@@ -91,7 +107,7 @@
     user = User.all.sample
     group = Group.all.sample
     if group.gm != user && !group.users.include?(user) && group.users.count < 5
-      p_group = PlayerGroup.create(
+      p_group = PlayerGroup.create!(
```
Comment (Ryu4fly) [inline db/seeds.rb:110]: Am thinking about putting this at the bottom of the seed and generate a Group that has users with the same preference in genre

Comment (Naokimi) [reply db/seeds.rb:110]: you could do something like
```
Genre.all.each do |genre|
  # users = players where genre: genre
  # groups = group where rulebook.genre: genre
  # assign players to these groups
end
```
alternatively you could first randomly assign players to groups, and then assign them preferences to match the groups they joined

**Comment by Naokimi** [comment]
@blamb888 @OoKiaoO 
merging this PR because Ryugi is taking a break and we can build on top of it

**Pull request merged**